SPEC 13: Recommended targets and naming conventions (#324)
scientific-python/specs: spec-0013/index.md

---
title: "SPEC 13 — Recommended targets and naming conventions"
number: 13
date: 2024-06-05
author:
  - "Pamphile Roy <roy.pamphile@gmail.com>"

>   - "Matthias Bussonnier <>"

Reviewer: ```suggestion
  - "Matthias Bussonnier <bussonniermatthias@gmail.com>"
```

  - "Jarrod Millman <millman@berkeley.edu>"
discussion: https://discuss.scientific-python.org/t/spec-13-recommended-targets-and-naming-conventions
endorsed-by:
---

## Description

> For consistency and decreased cognitive load across the ecosystem, this SPEC recommends naming conventions around various project aspects--such as project structure, repository layout, folder names, task runner and `pyproject.toml` targets name.

Reviewer 2: I would also add some defaults for declaring optional dependencies: `all`, `test`, and `docs`.

Reviewer 3: `test` -> `tests`? 🤔

Reviewer 2: never mind, should have read all the comments below before commenting, as this was already mentioned.

Reviewer 3: `dev` would also be nice.

Reviewer 2: test --> tests, bike shedding, but I'm 👎 on this. As it's test-dependencies in my view while the other recommendation is about a directory that contains the tests

Reviewer 3: Just hard to remember that "s" is for testing dir (`pytest tests/`) but no "s" for install (`pip install .[test]`).

Reviewer 2: astropy does it this way since forever 😄

Reviewer 4: Not too fond of `[dev]`, this makes people think it's okay to install all your development dependencies into a single environment, which is not true. Also not fond of `[all]`, same sort of reason. I've seen `[all]` used well for projects that have several optional dependencies and then provide an `[all]`, but it's not useful to add things like `[test]` dependencies.

Reviewer: We also limited the discussion to `test[s]` and `doc[s]` and did not discuss other target to now grow the spec too much. 

We can extend to other naming conventions later.

> From a cursory survey in the Scientific Python ecosystem, we discover some frustration from contributors and maintainer when moving from one project to another and belive that consistency will make it both easier for existing maintainer to contribute to manz project as well a decreae the confusion of new developers when contributing or creating new proejcts.

Reviewer 3: Typos
```suggestion
From a cursory survey in the Scientific Python ecosystem, we discover some frustration from contributors and maintainer when moving from one project to another and belive that consistency will make it both easier for existing maintainer to contribute to project as well a decrease the confusion of new developers when contributing or creating new projects.
```

## Implementation

We recommend that  by default the targets and folder names:
 - related to testing be named `tests` (and not `test`)
 - related to documentation be named `docs` (and not `doc`)
 - both should be lowercase.

It is ok to have the singular aliases, but the plurals should always be the default.

### Examples

> pyproj.toml

Reviewer: See https://learn.scientific-python.org/development/guides/packaging-simple/#extras

> pyproj.toml

Reviewer 3: ??

```suggestion
pyproject.toml
```

---
title: "SPEC 13 — Recommended targets and naming conventions"
number: 13
date: 2024-06-05
author:
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Matthias Bussonnier <bussonniermatthias@gmail.com>"
  - "Jarrod Millman <millman@berkeley.edu>"
discussion: https://discuss.scientific-python.org/t/spec-13-recommended-targets-and-naming-conventions
endorsed-by:
---

## Description

For consistency and decreased cognitive load across the ecosystem, this SPEC
recommends naming conventions around various project aspects--such as project
structure, repository layout, folder names, task runner and `pyproject.toml`
targets name.

> From a cursory survey in the Scientific Python ecosystem, we discover some
> frustration from contributors and maintainer when moving from one project to
> another and believe that consistency will make it both easier for existing
> maintainer to contribute to project as well a decrease the confusion of new
> developers when contributing or creating new projects.

Reviewer 5: "cursory" means hasty and without attention to detail, but suggests that it wasn't done well. Readers might not think it's appropriate to base a SPEC on a "cursory" survey. "Informal" is also true but may carry less of a negative connotation.

Isn't scientific python used in the lower case sense here?

```suggestion
From an informal of scientific Python developers, we discover some
frustration from contributors and maintainers when moving from one project to
another. We believe that consistency will make it both easier for existing
maintainers to contribute to projects as well as decrease the confusion of new
developers when creating or contributing to projects.
```
Or consider "creating new projects or contributing to existing ones". (I found it easy to misread "contributing or creating new".)

> There seem to be a strong consensus with preference for `docs` in favor of
> `docs`, and a preference for `tests` in favor of `test`.

Reviewer 5: ```suggestion
There seem to be a strong consensus with preference for `docs` over
`doc`, and a preference for `tests` over `test`.
```
"preference for tests" means that "tests" is preferred, but "in favor of test" means that "test" is preferred, so it was not clear to me which was preferred.

We will note though that the _extra_ optional dependencies on PyPI seem to favor
`test` (present on 7573 packages) vs `tests` (2362 times).

## Implementation

> For the tie being we will not pronounce ourselves on the optional extra
> `extra` dependency for `pyproject.toml`.

Reviewer 5: ```suggestion
For the time being, we will not specify a name of the optional extra
`extra` dependency for `pyproject.toml`.
```
And I'd suggest moving this to the end. It might be OK to precede this section with one about "Scope", but as-is, it looks like this is just one example of something that was not decided.

For other targets and folders we recommend that:

- Targets related to testing be named `tests` (and not `test`). For example
  `spin tests`, `python dev.py tests`, `nox -s tests`.

> - Folders containing tests be names `tests`.

Reviewer 5: ```suggestion
- Folders containing tests be named `tests`.
```

> - Targets related to documentations be named `docs` (and not `doc`). For example

Reviewer 5: ```suggestion
- Targets related to documentation be named `docs` (and not `doc`). For example
```
(It was mentioned during the discussion that we say "docs" even though "documentation" is the full word.)

  `spin docs`, `make docs`, `tox -s docs`.

> - That the documentation `extra` optional dependency be named `docs` (and not

Reviewer 2: I would like to think that adding a recommendation for `all` and `dev` are non-controversial, too?

Reviewer 4: I strongly recommend no recommendation on `[dev]`. Having an "all in one" environment makes the environment hard to solve, sometimes impossible to solve. Solvers are very sensitive to the number of dependencies in an environment. There was a time when you couldn't install `black` and `tensorflow` in the same environment due to dependency conflicts. I've seen solvers choke for minutes[^1] trying to solve updates on massive "all-in-one" environments. And you never need to install your code formatter, or pre-commit, etc. in the same environment as your package, they are stand-alone _tools_!

And `[all]` sometimes means all true optional dependencies, not testing and development dependencies. Like `validate-pyproject[all]`.

`uv` is currently working on a solution for project-level "pipx"-like installs. And there is an effort to make dependency groups a standard in [PEP 735](https://peps.python.org/pep-0735/). How about avoiding any extra recommendation until PEP 735 is accepted or rejected, and we look into seeing if there's interest in changing the official `[doc]` extra recommendation to `[docs]`?

[^1]: On average. The record I've seen was 25 hours.

Reviewer 2: OK, no strong preference for `[dev]`, but definitely having a true, user facing dependencies into `[all]`, so it should have no testing or documentation ones in there.

  doc), so that docs dependencies can be installed with `pip install .[docs]`
- Use lowercase.

> It is appropriate to have the singular aliases to ease transition, but the

Reviewer 5: Might mention what the aliases are for. Probably not aliases for folders?
